### PR TITLE
feat(plugin-chart-echarts): Echarts Waterfall

### PR DESCRIPTION
**Pull request by stephenLYZ**
### SUMMARY

Waterfall charts are useful for understanding how an initial value (like net income) is affected by a series of positive and negative changes.

For example, we can see the trend of personal annual income. Here we use the year as the series.
image
Moreover, we can use the value in breakdowns to add additional data to the chart. It adds the contributors to increases or decreases for each series.
image

### TESTING INSTRUCTIONS

1. Create a virtual dataset

<details><summary>Click me</summary>

```
SELECT * FROM 
(SELECT 'Facebook' as due_to_group, 2020 as period, 1945565 as value
UNION 
SELECT 'Competitor TV Advertising', 2019, 1213252
UNION 
SELECT 'Online Advertising', 2018, 999990
UNION 
SELECT 'COREBASE', 2017, 852094
UNION 
SELECT 'COREBASE', 2018, 736576
UNION
SELECT 'DISPLAY', 2017, 621608
UNION 
SELECT 'DISPLAY', 2018, 388904
UNION 
SELECT 'Facebook', 2019, 94909
UNION 
SELECT 'Online Advertising', 2017, 81334
UNION
SELECT 'Halo TV', 2018, 66828
UNION 
SELECT 'Halo TV', 2017, 46818
UNION 
SELECT 'Competitor TV Advertising', 2017, 25252
UNION 
SELECT 'Facebook', 2017, 23932
UNION
SELECT 'DFSI', 2017, 21466
UNION 
SELECT 'Coupons', 2017, 11160
UNION 
SELECT 'Facebook', 2018, 9444
UNION 
SELECT 'DFSI', 2019, 8785
UNION
SELECT 'Competitive Coupons', 2017, 8724
UNION 
SELECT 'Competitive Coupons', 2019, 8724
UNION 
SELECT 'Coupons', 2019, 2950
UNION 
SELECT 'BB Display', 2019, 1844
UNION
SELECT 'BB Display', 2017, 1844
UNION
SELECT 'Email', 2017, 810
UNION 
SELECT 'OTHER', 2017, 78
UNION 
SELECT 'Email', 2019, -987000
UNION 
SELECT 'Email', 2020, -998988
UNION 
SELECT 'Online Advertising', 2020, -1500000
UNION 
SELECT 'Online Advertising', 2019, -1671652
) as _view
```
</details>

3. Create a new chart by the dataset
4. Select "period" as "dimensions"
5. Select 'Sum(value)' as metric

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by cbuffevant** [comment]
👋 Is there any current effort in this PR? I'm really interested in this feature, and I'm more than happy to help

**Comment by rusackas** [comment]
I'm not sure if @stephenLYZ is still interested/willing to usher this through, so perhaps @yousoph might know someone who has the bandwidth to take it up?

**Comment by rusackas** [comment]
If you're volunteering to help finish this up, @cbuffevant, that would be more than welcome!!! The easiest way might be to start a new branch, cherry-pick these commits in (so @stephenLYZ gets credit for those) and keep piling on as needed. We could then close this PR in favor of the new one.

**Comment by stephenLYZ** [comment]
Sorry for being late.  I think we can merge it first and then add some improvement plans. Let me fix conflicts.

**Comment by cbuffevant** [comment]
Thanks @stephenLYZ ! I will be more than happy to help, but I've never contributed to Superset so I'm still not entirely familiar with the code...

**Comment by lucasqueiroz97** [comment]
Doing some tests here, if you insert this code at "types.ts": 

`export type WaterfallChartTransformedProps = BaseTransformedProps<EchartsWaterfallFormData> & CrossFilterTransformedProps;`

probably will work fine

**Comment by rusackas** [comment]
@cbuffevant if you're still able/willing to carry this forward, feel free to ping me on [slack](http://bit.ly/join-superset-slack) about getting that ball rolling :)

**Comment by michael-s-molina** [comment]
PR rebased

**Comment by rusackas** [comment]
OK, I got it working, but realized a few things along the way:
1) Storybook isn't running (not the fault of this PR, but just sayin'...)
2) This doesn't seem to really support a temporal X Axis via the "Dimension" input, which sure would be nice.
3) If the first values in a series are negative, they're still above the 0 line, and lead to a non-waterfall chart like so
Screenshot 2023-10-05 at 3 16 21 PM

That said, I was still able to get a reasonable working example. 
Screenshot 2023-10-05 at 4 01 21 PM

I say we merge this thing and then debug it on new PRs. I'll give the code a quick review...

File: superset-frontend/plugins/plugin-chart-echarts/src/Waterfall/index.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  constructor() {
+    super({
+      buildQuery,
+      controlPanel,
+      loadChart: () => import('./EchartsWaterfall'),
+      metadata: new ChartMetadata({
+        behaviors: [Behavior.INTERACTIVE_CHART, Behavior.DRILL_TO_DETAIL],
+        credits: ['https://echarts.apache.org'],
+        category: t('Evolution'),
+        description: '',
```
Comment (rusackas) [inline superset-frontend/plugins/plugin-chart-echarts/src/Waterfall/index.ts:50]: We might want to flesh this out in a subsequent PR.

Also, while I'm naming random follow-ups, I'll bet the farm that people will want conditional coloring, (e.g. upward is green, downward is red) but that's another non-blocking thing.

**Review by rusackas — APPROVED** [review]
There are improvements to make and things to fix, but I think we should merge this PR and "fix it in the mix" as they say.

**Pull request merged**